File: cpp/base/src/main/vsdk/toolkit/environment/geometry/volume/VoxelVolume.cpp
#include "vsdk/toolkit/environment/geometry/volume/VoxelVolume.h"
#include "vsdk/toolkit/media/IndexedColorImageUncompressed.h"
#include "vsdk/toolkit/environment/geometry/element/Ray.h"
#include "vsdk/toolkit/environment/geometry/element/RayHit.h"
#include "vsdk/toolkit/common/linealAlgebra/Matrix4x4d.h"
#include "vsdk/toolkit/common/VSDK.h"
#include <cmath>

VoxelVolume::VoxelVolume() : data(nullptr), xSize(0), ySize(0), zSize(0) {}

VoxelVolume::~VoxelVolume() {
    if (data != nullptr) {
        for (int z = 0; z < zSize; z++) {
            delete data[z];
        }
        delete[] data;
    }
}

int VoxelVolume::getXSize() const { return xSize; }
int VoxelVolume::getYSize() const { return ySize; }
int VoxelVolume::getZSize() const { return zSize; }

bool VoxelVolume::init(int xs, int ys, int zs) {
    if (xs <= 0 || ys <= 0 || zs <= 0) return false;

    if (data != nullptr) {
        for (int z = 0; z < zSize; z++) delete data[z];
        delete[] data;
        data = nullptr;
    }

    data = new IndexedColorImageUncompressed*[zs];
    xSize = xs; ySize = ys; zSize = zs;

    for (int z = 0; z < zSize; z++) {
        data[z] = new IndexedColorImageUncompressed();
        if (!data[z]->init(xSize, ySize)) {
            for (int i = 0; i <= z; i++) delete data[i];
            delete[] data;
            data = nullptr;
            xSize = ySize = zSize = 0;
            return false;
        }
    }
    return true;
}

void VoxelVolume::putVoxel(int x, int y, int z, char val) {
    if (data == nullptr) return;
    if (x < 0 || x >= xSize || y < 0 || y >= ySize || z < 0 || z >= zSize) return;
    data[z]->putPixel(x, y, val);
}

int VoxelVolume::getVoxel(int x, int y, int z) const {
    if (data == nullptr) return 0;
    if (x < 0 || x >= xSize || y < 0 || y >= ySize || z < 0 || z >= zSize) return 0;
    return (unsigned char)data[z]->getPixel(x, y);
}

Vector3Dd VoxelVolume::getVoxelPosition(int x, int y, int z) const {
    return Vector3Dd(
        (((double)x + 0.5) / (double)xSize) * 2 - 1,
        (((double)y + 0.5) / (double)ySize) * 2 - 1,
        (((double)z + 0.5) / (double)zSize) * 2 - 1);
}

int VoxelVolume::getNearestIFromX(double x) const { return (int)(((x + 1)/2) * ((double)xSize) - 0.5); }
int VoxelVolume::getNearestJFromY(double y) const { return (int)(((y + 1)/2) * ((double)ySize) - 0.5); }
int VoxelVolume::getNearestKFromZ(double z) const { return (int)(((z + 1)/2) * ((double)zSize) - 0.5); }

int VoxelVolume::getVoxelAtPosition(double x, double y, double z) const {
    if (x < -1 || x > 1 || y < -1 || y > 1 || z < -1 || z > 1) return 0;
    return getVoxel(getNearestIFromX(x), getNearestJFromY(y), getNearestKFromZ(z));
}

int VoxelVolume::getVoxelAtPosition(const Vector3Dd& p) const {
    return getVoxelAtPosition(p.x(), p.y(), p.z());
}

void VoxelVolume::putVoxelAtPosition(double x, double y, double z, char val) {
    if (x < -1 || x > 1 || y < -1 || y > 1 || z < -1 || z > 1) return;
    putVoxel(getNearestIFromX(x), getNearestJFromY(y), getNearestKFromZ(z), val);
}

void VoxelVolume::putVoxelAtPosition(const Vector3Dd& p, char val) {
    putVoxelAtPosition(p.x(), p.y(), p.z(), val);
}

double* VoxelVolume::getMinMax() {
    double* m = new double[6];
    m[0]=-1; m[1]=-1; m[2]=-1; m[3]=1; m[4]=1; m[5]=1;
    return m;
}

Ray* VoxelVolume::doIntersection(const Ray& inOutRay) {
    RayHit hit;
    if (doIntersection(inOutRay, &hit) && hit.ray() != nullptr) {
        return new Ray(*hit.ray());
    }
    return nullptr;
}

bool VoxelVolume::doIntersection(const Ray& inRay, RayHit* outHit) {
    double tmin = -1e308, tmax = 1e308;
    const double bounds[2] = {-1.0, 1.0};

    for (int axis = 0; axis < 3; axis++) {
        double o = axis == 0 ? inRay.getOrigin().x() : (axis == 1 ? inRay.getOrigin().y() : inRay.getOrigin().z());
        double d = axis == 0 ? inRay.getDirection().x() : (axis == 1 ? inRay.getDirection().y() : inRay.getDirection().z());

        if (std::abs(d) <= VSDK::EPSILON) {
            if (o < bounds[0] || o > bounds[1]) return false;
            continue;
        }

        double t1 = (bounds[0] - o) / d;
        double t2 = (bounds[1] - o) / d;
        if (t1 > t2) { double tmp = t1; t1 = t2; t2 = tmp; }
        if (t1 > tmin) tmin = t1;
        if (t2 < tmax) tmax = t2;
        if (tmin > tmax) return false;
    }

    double hitT = tmin > VSDK::EPSILON ? tmin : tmax;
    if (hitT <= VSDK::EPSILON) return false;

    if (outHit != nullptr) {
        if (outHit->shouldStoreRay() || outHit->needsAnySurfaceData()) outHit->setRay(inRay.withT(hitT));
        else outHit->setHitDistance(hitT);
        doExtraInformation(inRay, hitT, outHit);
    }
    return true;
}

void VoxelVolume::doExtraInformation(const Ray& inRay, double inT, RayHit* outData) {
    if (outData == nullptr) return;
    if (outData->needsPoint()) {
        outData->p = inRay.getOrigin().add(inRay.getDirection().multiply(inT));
    }
}

Matrix4x4d VoxelVolume::getTransformFromVoxelFrameToMinMax(const double minmax[6]) {
    double sx = minmax[3]-minmax[0], sy = minmax[4]-minmax[1], sz = minmax[5]-minmax[2];
    double greaterScale = sx;
    if (sy > greaterScale) greaterScale = sy;
    if (sz > greaterScale) greaterScale = sz;

    Matrix4x4d S, T1, T2;
    S = S.scale(greaterScale/2, greaterScale/2, greaterScale/2);
    T1 = T1.translation(1, 1, 1);
    T2 = T2.translation(minmax[0]-(greaterScale-sx)/2, minmax[1]-(greaterScale-sy)/2, minmax[2]-(greaterScale-sz)/2);
    return T2.multiply(S.multiply(T1));
}

Vector3Dd VoxelVolume::doCenterOfMass() {
    double cmx=0,cmy=0,cmz=0,mi,M=0;
    for (int x=0; x<getXSize(); x++) {
        for (int y=0; y<getYSize(); y++) {
            for (int z=0; z<getZSize(); z++) {
                mi = ((double)getVoxel(x,y,z))/255.0;
                M += mi;
                Vector3Dd p = getVoxelPosition(x,y,z);
                cmx += mi*p.x(); cmy += mi*p.y(); cmz += mi*p.z();
            }
        }
    }
    if (std::abs(M) < VSDK::EPSILON) return Vector3Dd(0,0,0);
    return Vector3Dd(cmx/M, cmy/M, cmz/M);
}

File: typescript/base/src/main/vsdk/toolkit/environment/geometry/volume/VoxelVolume.ts
//= References:                                                             =
//= [KAUF1987] Kaufman, Arie. "Efficient Algorithms for 3D Scan-Conversion  =
//=     of Parametric Curves, Surfaces, and Volumes", ACM SIGGRAPH Computer =
//=     Graphics, volume 21, number 4, July 1987.                           =

import { VSDK } from "../../../common/VSDK.js";
import type { Ray } from "../element/Ray.js";
import { Matrix4x4d } from "../../../common/linealAlgebra/Matrix4x4d.js";
import { Vector3Dd } from "../../../common/linealAlgebra/Vector3Dd.js";
import type { RayHit } from "../element/RayHit.js";
import { IndexedColorImageUncompressed } from "../../../media/IndexedColorImageUncompressed.js";
import { Solid } from "./Solid.js";

/**
VoxelVolume represents a voxelized paralelogram volume in memory, as the
"cubic frame buffer" proposed in [KAUF1987]. Note that this class is intended
for simpler applications in which data volume fix into main memory. Current
class doesn't support any caching or data storage optimization. When the
x/y/z sizes of the voxel volume are equal, the corresponding space covered
by the volume is the cube from the point <-1, -1, -1> to the point <1, 1, 1>.
The voxels are always assume to be square, and when dimensions are different,
the largest dimension fits in to the <-1, 1> interval (other dimensions are
proportional to maintain voxel sizes).

As current voxel volume is based in specific data samples of 8 bits per voxel,
it is not general and not well biased to processing applications. This is
just a placeholder for regions of interest in memory, as an aid to
other algorithms.

For general representation of N-dimensional images of arbitrary data sample
format, use another toolkit, like ITK or VTK.
*/
export class VoxelVolume extends Solid {
    private data: IndexedColorImageUncompressed[] | null;

    public constructor() {
        super();
        this.data = null;
    }

    public getXSize(): number {
        if (this.data === null || this.data.length < 0) return 0;
        return VoxelVolume.listGet(this.data, 0).getXSize();
    }

    public getYSize(): number {
        if (this.data === null || this.data.length < 0) return 0;
        return VoxelVolume.listGet(this.data, 0).getXSize();
    }

    public getZSize(): number {
        if (this.data === null || this.data.length < 0) return 0;
        return this.data.length;
    }

    public init(xSize: number, ySize: number, zSize: number): boolean {
        let z: number;
        let slice: IndexedColorImageUncompressed;

        const localData: IndexedColorImageUncompressed[] = [];

        for (z = 0; z < zSize; z++) {
            slice = new IndexedColorImageUncompressed();
            if (!slice.init(xSize, ySize)) {
                return false;
            }
            localData.push(slice);
        }

        this.data = localData;
        return true;
    }

    public putVoxel(x: number, y: number, z: number, val: number): void {
        try {
            if (x < 0 || x >= this.getXSize() || y < 0 || y >= this.getYSize() || z < 0 || z >= this.getZSize()) {
                return;
            }
            const slice = VoxelVolume.listGet(this.data!, z);
            // Java calls the `putPixel(int, int, byte)` overload.
            slice.putPixelByte(x, y, val);
        } catch {
            //
        }
    }

    /**
    Given current voxel set geometric space (cube from <-1, -1, -1> to
    <1, 1, 1>), current voxel set size, and cell position to a voxel; this
    methods gives the position of the voxel center in world coordinates.
    @param x
    @param y
    @param z
    @return a new Vector3Dd with point position corresponding to given indexes
    inside the matrix of voxels
    */
    public getVoxelPosition(x: number, y: number, z: number): Vector3Dd {
        let p = new Vector3Dd();
        p = p.withX(((x + 0.5) / this.getXSize()) * 2 - 1);
        p = p.withY(((y + 0.5) / this.getYSize()) * 2 - 1);
        p = p.withZ(((z + 0.5) / this.getZSize()) * 2 - 1);
        return p;
    }

    /**
    Partial coordinate convertion (X axis) for `x` voxel coordinate to
    corresponding voxel index.
    @param x
    @return a matrix index for the given coordinate
    */
    public getNearestIFromX(x: number): number {
        return VoxelVolume.toInt(((x + 1) / 2) * this.getXSize() - 0.5);
    }

    /**
    Partial coordinate convertion (Y axis) for `y` voxel coordinate to
    corresponding voxel index.
    @param y
    @return a matrix index for the given coordinate
    */
    public getNearestJFromY(y: number): number {
        return VoxelVolume.toInt(((y + 1) / 2) * this.getYSize() - 0.5);
    }

    /**
    Partial coordinate convertion (Z axis) for `z` voxel coordinate to
    corresponding voxel index.
    @param z
    @return a matrix index for the given coordinate
    */
    public getNearestKFromZ(z: number): number {
        return VoxelVolume.toInt(((z + 1) / 2) * this.getZSize() - 0.5);
    }

    /**
    Given current voxelset geometric space (cube from <-1, -1, -1> to
    <1, 1, 1>), current voxelset size, and cell position to a voxel; this
    methods gives the voxel value with a position corresponding to coordinate
    <x, y, z> (inside voxel space cube).

    Java overloads `getVoxelAtPosition(double, double, double)` and
    `getVoxelAtPosition(Vector3Dd)`.
    @return value of voxel at specified indexed position
    */
    public getVoxelAtPosition(p: Vector3Dd): number;
    public getVoxelAtPosition(x: number, y: number, z: number): number;
    public getVoxelAtPosition(a: Vector3Dd | number, b?: number, c?: number): number {
        let i: number;
        let j: number;
        let k: number;
        if (a instanceof Vector3Dd) {
            const p = a;
            if (p.x() < -1 || p.x() > 1 || p.y() < -1 || p.y() > 1 || p.z() < -1 || p.z() > 1) return 0;

            i = VoxelVolume.toInt(((p.x() + 1) / 2) * this.getXSize() - 0.5);
            j = VoxelVolume.toInt(((p.y() + 1) / 2) * this.getYSize() - 0.5);
            k = VoxelVolume.toInt(((p.z() + 1) / 2) * this.getZSize() - 0.5);

            return this.getVoxel(i, j, k);
        }
        const x = a;
        const y = b!;
        const z = c!;
        if (x < -1 || x > 1 || y < -1 || y > 1 || z < -1 || z > 1) return 0;

        i = VoxelVolume.toInt(((x + 1) / 2) * this.getXSize() - 0.5);
        j = VoxelVolume.toInt(((y + 1) / 2) * this.getYSize() - 0.5);
        k = VoxelVolume.toInt(((z + 1) / 2) * this.getZSize() - 0.5);

        return this.getVoxel(i, j, k);
    }

    /**
    Given current voxelset geometric space (cube from <-1, -1, -1> to
    <1, 1, 1>), current voxelset size, and cell position to a voxel; this
    methods puts the voxel value with a position corresponding to coordinate
    <x, y, z> (inside voxel space cube).

    Java overloads `putVoxelAtPosition(double, double, double, byte)` and
    `putVoxelAtPosition(Vector3Dd, byte)`.
    */
    public putVoxelAtPosition(p: Vector3Dd, val: number): void;
    public putVoxelAtPosition(x: number, y: number, z: number, val: number): void;
    public putVoxelAtPosition(a: Vector3Dd | number, b: number, c?: number, d?: number): void {
        let i: number;
        let j: number;
        let k: number;
        if (a instanceof Vector3Dd) {
            const p = a;
            const val = b;
            if (p.x() < -1 || p.x() > 1 || p.y() < -1 || p.y() > 1 || p.z() < -1 || p.z() > 1) return;

            i = VoxelVolume.toInt(((p.x() + 1) / 2) * this.getXSize() - 0.5);
            j = VoxelVolume.toInt(((p.y() + 1) / 2) * this.getYSize() - 0.5);
            k = VoxelVolume.toInt(((p.z() + 1) / 2) * this.getZSize() - 0.5);

            this.putVoxel(i, j, k, val);
            return;
        }
        const x = a;
        const y = b;
        const z = c!;
        const val = d!;
        if (x < -1 || x > 1 || y < -1 || y > 1 || z < -1 || z > 1) return;

        i = VoxelVolume.toInt(((x + 1) / 2) * this.getXSize() - 0.5);
        j = VoxelVolume.toInt(((y + 1) / 2) * this.getYSize() - 0.5);
        k = VoxelVolume.toInt(((z + 1) / 2) * this.getZSize() - 0.5);

        this.putVoxel(i, j, k, val);
    }

    public getVoxel(x: number, y: number, z: number): number {
        try {
            const slice = VoxelVolume.listGet(this.data!, z);
            // Java's backing byte[] throws ArrayIndexOutOfBoundsException for
            // an index outside the slice, which this method catches; the
            // TypeScript typed array returns undefined instead, so the same
            // bound is checked explicitly.
            const index = slice.getXSize() * y + x;
            if (index < 0 || index >= slice.getXSize() * slice.getYSize()) {
                throw new RangeError("Array index out of range: " + index);
            }
            return slice.getPixel(x, y);
        } catch {
            //
        }
        return 0;
    }

    /**
    Check the general interface contract in superclass method
    Geometry.getMinMax.
    @return a new 6 valued double array containing the coordinates of a min-max
    bounding box for current geometry.
    */
    public override getMinMax(): Float64Array {
        const minMax = new Float64Array(6);
        minMax[0] = -1.0;
        minMax[1] = -1.0;
        minMax[2] = -1.0;
        minMax[3] = 1.0;
        minMax[4] = 1.0;
        minMax[5] = 1.0;
        return minMax;
    }

    /**
    Check the general interface contract in superclass method
    Geometry.doIntersectionFirstHit.

    NOT IMPLEMENTED YET!

    Java overload `doIntersectionFirstHit(Ray inOut_Ray)` returns null.
    @return true if given ray intersects current VoxelVolume
    */
    public doIntersectionFirstHit(inOut_Ray: Ray): Ray | null;
    public override doIntersectionFirstHit(inRay: Ray, outHit: RayHit): boolean;
    public override doIntersectionFirstHit(inRay: Ray, outHit: RayHit | null): boolean;
    public override doIntersectionFirstHit(_inRay: Ray, outHit?: RayHit | null): Ray | null | boolean {
        if (outHit === undefined) {
            return null;
        }
        return false;
    }

    /**
    Check the general interface contract in superclass method
    Geometry.doExtraInformation.
    @param inRay
    @param inT
    @param outData
    */
    public override doExtraInformation(_inRay: Ray, _inT: number, _outData: RayHit): void {}

    /**
    Current method creates a transformation matrix that represent the
    coordinate change from voxel volume cube <-1, -1, -1>-<1, 1, 1> to
    the bounding box recieved in `minmax`.
    @param minmax
    @return a new Matrix4x4d for coordinate mapping between volume and
    world coordinates
    */
    public static getTransformFromVoxelFrameToMinMax(minmax: Float64Array | readonly number[]): Matrix4x4d {
        let S: Matrix4x4d;
        let T1: Matrix4x4d;
        let T2: Matrix4x4d;
        let greaterScale: number;

        const sx = minmax[3]! - minmax[0]!;
        const sy = minmax[4]! - minmax[1]!;
        const sz = minmax[5]! - minmax[2]!;
        greaterScale = sx;
        if (sy > greaterScale) {
            greaterScale = sy;
        }
        if (sz > greaterScale) {
            greaterScale = sz;
        }

        S = new Matrix4x4d();
        S = S.scale(greaterScale / 2, greaterScale / 2, greaterScale / 2);

        T1 = new Matrix4x4d();
        T1 = T1.translation(1, 1, 1);
        T2 = new Matrix4x4d();
        T2 = T2.translation(
            minmax[0]! - (greaterScale - sx) / 2,
            minmax[1]! - (greaterScale - sy) / 2,
            minmax[2]! - (greaterScale - sz) / 2,
        );

        const M = T2.multiply(S.multiply(T1));
        return M;
    }

    /**
    Check the general interface contract in superclass method
    Solid.doCenterOfMass
    @return new Vector3Dd containing volume center of mass
    */
    public override doCenterOfMass(): Vector3Dd {
        let p: Vector3Dd;
        let cmx = 0;
        let cmy = 0;
        let cmz = 0;
        let mi: number; // Maximum mass of one voxel (linear to voxel density)
        let M = 0; // Total mass for current voxel volume
        let x: number;
        let y: number;
        let z: number;

        for (x = 0; x < this.getXSize(); x++) {
            for (y = 0; y < this.getYSize(); y++) {
                for (z = 0; z < this.getZSize(); z++) {
                    // mi goes from 0 to 1
                    mi = this.getVoxel(x, y, z) / 255.0;
                    M += mi;
                    p = this.getVoxelPosition(x, y, z);
                    cmx += mi * p.x();
                    cmy += mi * p.y();
                    cmz += mi * p.z();
                }
            }
        }

        if (Math.abs(M) < VSDK.EPSILON) {
            return new Vector3Dd(0, 0, 0);
        }

        return new Vector3Dd(cmx / M, cmy / M, cmz / M);
    }

    /** Java `ArrayList.get`, which throws for an index outside the list. */
    private static listGet(
        list: readonly IndexedColorImageUncompressed[],
        index: number,
    ): IndexedColorImageUncompressed {
        if (index < 0 || index >= list.length) {
            throw new RangeError("Index " + index + " out of bounds for length " + list.length);
        }
        return list[index]!;
    }

    /** Java narrowing conversion `(int)double`. */
    private static toInt(value: number): number {
        if (Number.isNaN(value)) return 0;
        if (value >= 2147483647) return 2147483647;
        if (value <= -2147483648) return -2147483648;
        return Math.trunc(value);
    }
}
